Спринт 11

ZeroXKiritsu opened the pull request.

Hardwired8403 committed:
Спринт 11

Hardwired8403 committed:
fix: delete unused imports

Hardwired8403 committed:
fix:  WhitespaceAround: '{' is not preceded with whitespace.

Hardwired8403 committed:
feat: Добавил валидатор кастомный

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -15,36 +16,37 @@
 @RestController
 @RequestMapping(value = "/films", produces = "application/json")
 public class FilmController {
+    private final FilmStorage filmStorage;
```
Контроллер обращается к сервису, а сервис обращается к репозиторию. То есть в контроллере не должно ничего знать про репозиторий.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -18,18 +19,18 @@
 
     private final UserService userService;
-
+    private final UserStorage userStorage;
```
Аналогично.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,146 @@
+                    .duration(filmRows.getInt("duration"))
+                    .mpa(mpaDbStorage.getMpa(filmRows.getInt("rating_mpa_id")))
+                    .build();
+            film.setGenres(genreDbStorage.getGenreForCurrentFilm(film.getId()));
```
Не очень хорошо делать запросы к БД в цикле. Это большая нагрузка на БД. Затратно по времени из-за передачи по сети.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,91 @@
+        }
+        film.getGenres().forEach(g -> {
+            String sqlQuery = "INSERT INTO genre(film_id, genre_id) VALUES (?, ?)";
+            jdbcTemplate.update(sqlQuery,
```
Запросы к БД в цикле.

sergeymamonov commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -24,44 +26,62 @@ public class FilmControllerTest {
     void createFilm_IncorrectDescription_badRequestTest() {
-        testFilm.setDescription("a".repeat(201));
-        Assertions.assertThrows(ValidationException.class, () -> controller.create(testFilm), "Превышено количество символов в описании фильма.");
+        testFilm.setDescription("Размер описания значительно превышает двести символов, а может и не превышает " +
```
Если так делать строку определенной длины, то это сложно и легко ошибиться. Плюс непонятно, что этой строкой проверяется.
Вместо этого "a".repeat(нужное количество раз) - так сразу понятно, мы проверяем принимаемое полем количество символов.

sergeymamonov reviewed (commented):
ErrorResponse - можно воспользоваться аннотациями ломбока чтобы не писать конструктор и геттер.

Hardwired8403 committed:
refactor: Update FilmController and UserController to use service layer

- Updated FilmController to use FilmService for adding, updating, and retrieving films.
- Updated UserController to use UserService for adding, updating, and retrieving users.
- Removed FilmStorage and UserStorage dependencies from FilmController and UserController.
- Updated the import statements in both controllers to remove unused references.
- Updated ErrorResponse model class to use lombok annotations for constructor and getter methods.
- Updated FilmService to delegate operations to FilmStorage for adding, updating, and retrieving films.

Hardwired8403 committed:
Refactor FilmController and UserController constructors

The constructors for the FilmController and UserController classes have been refactored to remove redundant dependencies. The `FilmController` constructor no longer requires `filmStorage` and the `UserController` constructor no longer requires `userStorage`. This change simplifies the constructors and improves code readability.

Refactors:
- `FilmController`: Removed `filmStorage` parameter in constructor
- `UserController`: Removed `userStorage` parameter in constructor

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,163 @@
+                "FROM films WHERE film_id=?";
+        try {
+            return jdbcTemplate.queryForObject(sqlQuery, this::mapRowToFilm, id);
+        } catch (RuntimeException e) {
```
Если перехватывать все ошибки (RuntimeException), то в случае возникновения NPE или сеть к БД отвалилась, пользователю вернется не ошибка приложения, а NotFoundException(" Фильи не найден"). Нужно перехватывать именно ту ошибку(ошибки), которую кидает метод, который обернут в try/catch/

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,95 @@
+        String sqlQuery = "SELECT genre_id, name FROM genre_type WHERE genre_id=?";
+        try {
+            return jdbcTemplate.queryForObject(sqlQuery, this::mapRowToGenre, id);
+        } catch (RuntimeException e) {
```
Аналогично.

sergeymamonov reviewed (commented):
Везде, где перехватывается RuntimeException нужно изменить на перехват именно той ошибки, которую кидает метод в блоке try. RuntimeException - это необрабатываемое исключение, поэтому его в данных случаях перехватывать не нужно.

Hardwired8403 committed:
Fix exception handling in FilmDbStorage and GenreDbStorage

The modification involves updating the exception handling in the `FilmDbStorage` and `GenreDbStorage` classes. In both classes, the catch block is changed to catch `DataAccessException` instead of `RuntimeException`. This change ensures that the correct exception type is caught and helps improve code readability and maintainability.

This commit fixes the exception handling in the FilmDbStorage and GenreDbStorage classes. Now, they catch `DataAccessException` instead of `RuntimeException`. This change ensures that the proper exception type is caught and improves the code's robustness.

Please note that this commit message does not provide a detailed explanation or implementation rationale, as requested.
